tree/ntuple/v7/src/RNTupleProcessor.cxx

```diff
@@ -116,3 +185,188 @@ ROOT::Experimental::NTupleSize_t ROOT::Experimental::RNTupleChainProcessor::Adva
+
+      // Skip if the field is the untyped record that holds the fields of auxiliary ntuples.
+      if (fAuxiliaryPageSources.find(field.GetFieldName()) != fAuxiliaryPageSources.end())
+         continue;
+
+      if (fFieldContexts.find(field.GetQualifiedFieldName()) == fFieldContexts.end()) {
+         // The field belongs to the auxiliary ntuple currently being added.
+         // Apart from assigning its entry value the correct pointer, we also have to create a field context for it.
+
+         // If the model has a default entry, use the value pointers from the entry in the entry managed by the
+         // processor. This way, the pointers returned by RNTupleModel::MakeField can be used in the processor loop to
+         // access the corresponding field values.
+         if (!model->IsBare()) {
+            auto valuePtr = model->GetDefaultEntry().GetPtr<void>(field.GetFieldName());
+            newEntry->BindValue(field.GetQualifiedFieldName(), valuePtr);
+         }
+
+         auto token = newEntry->GetToken(field.GetQualifiedFieldName());
+         fFieldContexts.try_emplace(field.GetQualifiedFieldName(), field.Clone(field.GetFieldName()), token,
+                                    auxNTuple.fNTupleName, true /* isAuxiliary */);
+      } else {
+         auto valuePtr = fEntry->GetPtr<void>(field.GetQualifiedFieldName());
+         newEntry->BindValue(field.GetQualifiedFieldName(), valuePtr);
+      }
+   }
+
+   fEntry.swap(newEntry);
+
+   if (fUseIndex) {
+      fJoinIndices.try_emplace(auxNTuple.fNTupleName,
+                               Internal::RNTupleIndex::Create(fJoinFieldNames, pageSource, true /* deferBuild */));
+   }
+}
+
+void ROOT::Experimental::RNTupleJoinProcessor::ConnectFields()
+{
+   for (auto &[_, fieldContext] : fFieldContexts) {
+      Internal::RPageSource &pageSource =
+         fieldContext.IsAuxiliary() ? *fAuxiliaryPageSources.at(fieldContext.GetNTupleName()) : *fPageSource;
+      ConnectField(fieldContext, pageSource, *fEntry);
+   }
+}
+
+ROOT::Experimental::NTupleSize_t ROOT::Experimental::RNTupleJoinProcessor::Advance()
+{
+   ++fNEntriesProcessed;
+
+   if (fNEntriesProcessed >= fPageSource->GetNEntries()) {
+      return kInvalidNTupleIndex;
+   }
+
+   ++fLocalEntryNumber;
+   LoadEntry();
+
+   return fNEntriesProcessed;
+}
+
+void ROOT::Experimental::RNTupleJoinProcessor::LoadEntry()
+{
+   std::vector<void *> valPtrs;
+   valPtrs.reserve(fJoinFieldNames.size());
+
+   for (const auto &fieldName : fJoinFieldNames) {
+      auto ptr = fEntry->GetPtr<void>(fieldName);
+      valPtrs.push_back(ptr.get());
+   }
+
+   for (auto &[fieldName, fieldContext] : fFieldContexts) {
+      if (!fieldContext.fIsAuxiliary || !fUseIndex) {
+         auto &value = fEntry->GetValue(fieldName);
+         value.Read(fLocalEntryNumber);
+         continue;
+      }
+
+      auto joinIndex = fJoinIndices.find(fieldContext.GetNTupleName());
+      if (joinIndex != fJoinIndices.end()) {
```

Comment:
Can this ever happen? In case of `fUseIndex`, this looks like a failure case to me...